### DISCUSSION TITLE
How can I optimize the generator every five batches and optimize the discriminator every batch in GAN?

### DISCUSSION THREAD

**Discussion by Luciennnnnnn** [Lightning-AI/pytorch-lightning#11626] [code help: CV]
The [solution](https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#step-optimizeres-at-arbitrary-intervals) listed in the doc seems wasteful since when there is no need to update the discriminator, the training_step / loss.backward() still be called.
I know I can use manual optimization, but how can I achieve this based on automatic optimization?

**Comment by akihironitta** [comment]
@LuoXin-s Apparently, users are required to call the closure function (which calls forward, zero_grad, backward) inside `LightningModule.optimizer_step`. It used to be optional, but #9360 made it required for closure to be called.

Maybe, we should make the closure call optional in the hook, e.g., change `MisconfigurationException` to some warning as commented in https://github.com/PyTorchLightning/pytorch-lightning/pull/9360#discussion_r714184387?

cc @carmocca (author of #9360)

**Reply by carmocca** [reply]
I still think that forcing it to be required has greater benefits than disadvantages as not calling it was a common user error.

What's the problem with using manual optimization here?
